Add MP OpenAPI UI OAuth Selenium Tests

abutch3r opened the pull request:
Covers both default and customized endpoints and goes through the full login flow

Update basic UI test to also test HTTPS in case something does happen there and matches the existing openapi-3.1 selenium test cases

Fixes #25437

abutch3r committed:
Add MP OpenAPI UI OAuth Selenium Tests

Covers both default and customized behavior and goes through the full login flow

Update basic UI test to also test HTTPS in case something does happen there and matches the existing openapi-3.1 selenium test cases

abutch3r committed:
Add MP OpenAPI UI OAuth Selenium Tests

Covers both default and customized behavior and goes through the full login flow

Update basic UI test to also test HTTPS in case something does happen there and matches the existing openapi-3.1 selenium test cases

Fully close browser at end of OpenAPI Selenium tests

Reduces potential chances of a browser connection preventing the stopping of the server.

Remove debug for PausableComponentQuiesceListener - only one comeponent defaultHttpEndpoint implements it.

abutch3r commented:
#build

abutch3r committed:
Close browser when test completes to stop any open connnections

Add debug for Server failing to stop and make sure server start is clean

abutch3r commented:
#build

abutch3r committed:
Merge remote-tracking branch 'personal/openapi-oauth-test' into openapi-oauth-test

# Conflicts:
#	dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIBasicTest.java
#	dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java

abutch3r commented:
#build #libby

abutch3r commented:
#build

abutch3r committed:
Remove debug from Pausable listenener

Tidy up comments

abutch3r commented:
#libby #build

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/bnd.bnd:
```diff
@@ -20,11 +20,20 @@ fat.project: true
-	io.openliberty.jakarta.restfulWS.3.0;version=latest
+	io.openliberty.jakarta.restfulWS.3.0;version=latest,\
+    io.openliberty.org.eclipse.microprofile.openapi.3.1;version=latest,\
+    jakarta.annotation-api;version=latest
```
Indentation inconsistent

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        WebElement title = waitForElement(driver, By.cssSelector("h2.title"), LONG_WAIT);
+        assertThat("Page title", title.getText(), Matchers.containsString("Generated API"));
+
+        //check that the operation is "unlocked"
+        //we will use the lock object later to check that it has been updated
+        WebElement lock = waitForElement(driver, By.cssSelector("button.authorization__btn.unlocked"));
```
The CSS selector seems a bit general here. If this is the lock button on the operation, it might be better to find the operation div first and then search within that.

`waitForElement` should work with either `driver` or with a root element to search under.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        //assertThat("Authorize button", authorizeButton.findElement(By.cssSelector("span")).getText(), Matchers.containsString("Authorize"));
+        authorizeButton.click();
+
+        //check that the modal loads
+        WebElement oauthModal = waitForElement(driver, By.cssSelector("div.modal-ux-header"));
+        WebElement oauthModalHeader = oauthModal.findElement(By.cssSelector("h3"));
+        assertThat("", oauthModalHeader.getText(),Matchers.containsString("Available authorizations"));
+
+        // Get the OAuth Authentication container
+        //WebElement authContainer = oauthModal.findElement(By.cssSelector("div.auth-container"));
+        WebElement authContainer = waitForElement(driver, By.cssSelector("div.auth-container"));
+        assertThat("Check that we ",authContainer.findElement(By.cssSelector("h4")).getText(),Matchers.containsString("oauth (OAuth2, authorizationCode)"));
```
Assertion messages look incomplete here?

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+
+        //Check the Authorize button is available and click it
+        WebElement authorizeButton = waitForElement(driver, By.xpath("//span[contains(.,'Authorize')]"));
+        //assertThat("Authorize button", authorizeButton.findElement(By.cssSelector("span")).getText(), Matchers.containsString("Authorize"));
```
Commented out code

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        assertThat("", oauthModalHeader.getText(),Matchers.containsString("Available authorizations"));
+
+        // Get the OAuth Authentication container
+        //WebElement authContainer = oauthModal.findElement(By.cssSelector("div.auth-container"));
+        WebElement authContainer = waitForElement(driver, By.cssSelector("div.auth-container"));
```
More commented out code. Was the `auth-container` not found under the modal or did it take a moment to appear? You should be able to pass `oauthModel` to `waitForElement` if the latter case.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        WebElement clientSecretField = authContainer.findElement(By.id("client_secret"));
+        clientSecretField.sendKeys(OAUTH_CLIENT_SECRET);
+        // Tick the scope tick box
+        //WebElement scopeCheckbox = authContainer.findElement(By.id("test-authorizationCode-checkbox-oauth"));
+        authContainer.findElement(By.cssSelector("span.item")).click();
```
`span.item` again seems pretty general

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        //This opens a new tab in the browser which Selenium does not yet know about and it itself is not focused on so we need to switch to it
+        authBtn.click();
+
+        //Switch to new tab
+        new WebDriverWait(driver,Duration.ofSeconds(3)).until(driver -> driver.getWindowHandles().size() == 2);
+        for(String windowId:driver.getWindowHandles()){
+            //as we already have the first tabs handle we go to the other one
+            if(!windowId.equals(originalWindow)){
+                driver.switchTo().window(windowId);
+            }
+        }
```
This code looks good but the file needs formatting.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIOauthTest.java:
```diff
@@ -0,0 +1,226 @@
+        //check button text has changed to Logout
+        assertThat("Auth button text should now state logout",authBtn.getText().equals("Logout"));
+
+        //close Authorization modal
+        waitForElement(driver,By.cssSelector("button.close-modal")).click();
```
This should maybe be searching under the modal.

Azquelt reviewed (commented):
This is looking really good, just a few bits to tidy up.

abutch3r committed:
Update test to improve accuracy of selections

Do a better job of building paths between elements to reduce potential false hits and improve accuracy when finding certain elements

Fix bnd formatting

abutch3r commented:
#libby #build

abutch3r commented:
#build

Azquelt reviewed (approved):
Great job, it's really good to have proper tests of the OAuth flow.

abutch3r commented:
#build

abutch3r commented:
#build
